### PR TITLE
Meter string conversion of interpreter values

### PR DESCRIPTION
**Pull request by SupunS**
Work towards https://github.com/dapperlabs/cadence-private-issues/issues/36

## Description

Introduced a `ToMeteredString()` method to be used only when string conversion is performed during the interpreter/runtime, where metering is needed.

Reduces the complexity of having to update all places where `RecursiveString()`/`String()` method is called, but metering is not needed. e.g: tooling, etc.
______

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: runtime/interpreter/function.go [hunk 1]
```diff
@@ -101,6 +102,12 @@ func (f *InterpretedFunctionValue) RecursiveString(_ SeenReferences) string {
 	return f.String()
 }
 
+func (f *InterpretedFunctionValue) ToMeteredString(memoryGauge common.MemoryGauge, _ SeenReferences) string {
+	typeString := f.Type.String()
```
Comment (SupunS) [inline runtime/interpreter/function.go:106]: We need to separately meter string conversions of runtime types. I'll open a subsequent PR for that

**Review by dsainati1 — APPROVED** [review]
Nice! This is a much better solution than requiring that `String` and `RecursiveString` have access to an interpreter.

**Review by robert-e-davidson3 — APPROVED** [review]

File: runtime/format/string.go [hunk 2]
```diff
@@ -25,3 +25,28 @@ import (
...
+			// ASCII non-printable characters (i.e: out of range from space through DEL-1)
+			if 0x20 > r || r > 0x7E {
+				asciiNonPrintableChars++
+			}
+		}
+	}
+
+	// len = printableChars + (escapedChars x 2) + (nonPrintableChars x 8) + (quote x 2)
+	l := (len(s) - asciiNonPrintableChars - escapedChars) + escapedChars*2 + asciiNonPrintableChars*8 + 2
+	return l
```
Comment (turbolent) [inline runtime/format/string.go:51]:
```suggestion
	return (len(s) - asciiNonPrintableChars - escapedChars) + escapedChars*2 + asciiNonPrintableChars*8 + 2
```

File: runtime/interpreter/account.go [hunk 3]
```diff
@@ -111,9 +112,11 @@ func NewAuthAccountValue(
 	}
 
 	var str string
-	stringer := func(_ SeenReferences) string {
+	stringer := func(memoryGauge common.MemoryGauge, _ SeenReferences) string {
 		if str == "" {
-			str = fmt.Sprintf("AuthAccount(%s)", address)
+			common.UseMemory(memoryGauge, common.NewRawStringMemoryUsage(13))
```
Comment (turbolent) [inline runtime/interpreter/account.go:117]: Maybe add this to the list of known memory usages up top. Same for `PublicAccount`, `AuthAccount.Contracts`, etc below

**Review by turbolent — COMMENTED** [review]
Sorry for the late review, missed this

File: runtime/interpreter/account.go (hunk 3 above)
Comment (SupunS) [reply runtime/interpreter/account.go]: Good idea! Added with 0b763c8a88be1f222c7e286b24f15a6a00db0f19

**Review by turbolent — APPROVED** [review]
Nice!

**Pull request merged**